### ISSUE TITLE
Windows: "bash: git: command not found" error thrown in RStudio shell, but not in Git Bash

### ISSUE THREAD

**Issue by everetr** [jennybc/happy-git-with-r#67]
**tl;dr,** In the book, can you consider telling Windows users to select "Use Git from the Windows Command Prompt" when installing Git? I believe it prevents the following headache.

*****

I ran into the error described [here](https://stackoverflow.com/questions/45665958/terminal-window-cannot-find-git-executable-rstudio-1-1-331). My scenario:

* Windows 10 (64-bit). Had latest versions of R ("short summer") and RStudio (1.1.383).
* Had latest version of Git from [here](https://git-scm.com/downloads). 
* Running `which git` in Git Bash returned the `git.exe` location. 
* I was able to run all `git` commands in Git Bash with no errors. Could set username and email.
* I then opened an existing R project already under version control in RStudio ("Git" menu was visible). 
* Running `which git` gave a really long error message.
* Running `git --version` and `git config --global --list` threw the error, `bash: git: command not found`.
* I confirmed that "Version control system:" in *Tools > Project Options > Git/SVN* was set to "Git".
* I confirmed that "Git executable" in *Tools > Global Options > Git/SVN* was set to the same exact path as shown by running `which git` in Git Bash. 
* ~~I confirmed that the “Use Git Bash as shell for Git projects” box in *Tools > Global Options > Git/SVN* was ticked.~~ **Edit:** This option is not in RStudio v1.1.383. My confusion.
* Name of the Rstudio git shell window was `MSYS`, not `cmd.exe` nor `MINGW64`, as you warned in the book [here](http://happygitwithr.com/hello-git.html).

My resolution:
* Re-installed git from [here](https://git-scm.com/downloads) (did not uninstall).
* During installation, selected "Use Git from the Windows Command Prompt" (different from default). Left all other options throughout the installation at their default settings.
* Running `git config --global --list` in Git Bash showed my same username and email. No need to reset.
* Running `which git` in RStudio's shell returned the `git.exe` location.
* All `git` commands work in RStudio's shell now.
* Am able to set username and email in RStudio's shell.

Perhaps this is worth noting in the book? A good spot would be under section 8.1 [here](http://happygitwithr.com/hello-git.html), since related notes for Windows users are already there.

**Comment by jennybc** [comment]
Would you like to make a PR for this? This is the kind of practical detail I definitely want to capture here and Windows is not my OS.

cc @daattali @kevinushey (does this all sound right to you?)

**Comment by daattali** [comment]
Definitely really useful to have there. I usually love helping you but I'm out of commission on non essential items for the next few weeks, sorry!

:hankey:
:tada: 

( ^ sorry to be a party pooper)

**Comment by kevinushey** [comment]
@everetr: Can you confirm that you are indeed running with RStudio v1.1.383? Normally, the git shell itself places extra utilities on the PATH. Do you by any chance have a startup profile / script that modifies the PATH? For posterity, I see:

<img width="349" alt="screen shot 2017-11-03 at 4 08 00 pm" src="https://user-images.githubusercontent.com/1976582/32399177-3c7517ea-c0b1-11e7-8ac1-bc56da752a73.png">

In addition, this:

> I confirmed that the “Use Git Bash as shell for Git projects” box in Tools > Global Options > Git/SVN was ticked.

We actually removed that option from v1.1.383, as users found it confusing that the type of shell launched differed between Git projects and non-Git projects. Now, we always launch a shell matching the one selected in the `Tools -> Global Options -> Terminal` pane, e.g.

<img width="475" alt="screen shot 2017-11-03 at 4 10 25 pm" src="https://user-images.githubusercontent.com/1976582/32399220-8647604e-c0b1-11e7-95fd-9d58ae0e537a.png">

I'd also be curious to see what the output of `echo $PATH | tr ':' '\n'` is in the shell launched by RStudio; e.g. I have:

```
kevin@DESKTOP-KLBFU53 MSYS /c/Users/kevin/Documents
$ echo $PATH | tr ':' '\n'
/c/Users/kevin/Documents/=/bin
/usr/local/bin
/usr/bin
/bin
/opt/bin
/c/R/R-3.4.2patched/bin/i386
/c/ProgramData/Oracle/Java/javapath
/c/WINDOWS/system32
/c/WINDOWS
/c/WINDOWS/System32/Wbem
/c/WINDOWS/System32/WindowsPowerShell/v1.0
/c/Program Files/MiKTeX 2.9/miktex/bin/x64
/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit
/c/Users/kevin/AppData/Local/Microsoft/WindowsApps
/c/Program Files/CMake/bin
/c/Program Files (x86)/WinAnt/bin
/cmd
/c/Program Files/RStudio/bin/msys-ssh-1000-18
/usr/bin/vendor_perl
/usr/bin/core_perl
```

Note the entry called `/cmd`.

**Comment by everetr** [comment]
@kevinushey 

My mistake. “Use Git Bash as shell for Git projects” is in section 8.1 of the book. I did NOT confirm that that box was checked, because it's not in my RStudio. 

I'll try to recreate this error and send you screenshots and the output of `echo $PATH | tr ':' '\n'`

**Comment by krshedd** [comment]
I was updating my version of RStudio and Git today and had the exact same issue described by @everetr. His resolution (re-installed with "Use Git from the Windows Command Prompt") worked for me as well. Thanks for posting.

**Comment by jennybc** [comment]
Ha! I set up a Windows VM last night and made sure to select this myself. It was selected by default for me. But it sounds like we need to repeat or move this advice, yet?

**Comment by everetr** [comment]
Just re-download Git (did not uninstall) and reproduced the error described in OP. 

## Installation settings:

![image](https://user-images.githubusercontent.com/9023041/32707833-a7068202-c7f4-11e7-889c-63d6732691a2.png)

![image](https://user-images.githubusercontent.com/9023041/32707825-9465930e-c7f4-11e7-9266-178102cbfc67.png)

![image](https://user-images.githubusercontent.com/9023041/32707846-b8e5f728-c7f4-11e7-8c70-e50565adebcf.png)

![image](https://user-images.githubusercontent.com/9023041/32707858-cd814b24-c7f4-11e7-8929-1465545bb266.png)

![image](https://user-images.githubusercontent.com/9023041/32707867-da31d1ae-c7f4-11e7-83f4-88c4a993bafb.png)

![image](https://user-images.githubusercontent.com/9023041/32707873-e771dfe4-c7f4-11e7-8c2d-35477d97add8.png)

## Output of `which git` in Git Bash:

![image](https://user-images.githubusercontent.com/9023041/32707895-10a91ddc-c7f5-11e7-890e-f9283766d426.png)

## Output of `which git` in RStudio Shell:

![image](https://user-images.githubusercontent.com/9023041/32708017-acc82366-c7f5-11e7-8e7e-97e5db364056.png)

## Output of `git --version` in Git Bash:

![image](https://user-images.githubusercontent.com/9023041/32708126-4527f096-c7f6-11e7-95eb-6d9d4becaf59.png)

## Output of `git --version` in RStudio Shell:

![image](https://user-images.githubusercontent.com/9023041/32708106-30529e32-c7f6-11e7-84cb-126e466a5bd4.png)

## Output of `git config --global --list` in Git Bash:

![image](https://user-images.githubusercontent.com/9023041/32708195-9d210f58-c7f6-11e7-99f7-774e4e54fedd.png)

## Output of `git config --global --list` in RStudio Shell:

![image](https://user-images.githubusercontent.com/9023041/32708150-694ca390-c7f6-11e7-9010-dd82f06bab6f.png)

## RStudio Version:

![image](https://user-images.githubusercontent.com/9023041/32708048-e3d4005a-c7f5-11e7-9275-c680955cf534.png)

*****

Git path set correctly in Global Options.

*****

## Output of @kevinushey's requested command in RStudio Shell:

![image](https://user-images.githubusercontent.com/9023041/32708243-e0017c18-c7f6-11e7-9d57-48af56708ca9.png)

# Reinstalling, the only option I change is:

![image](https://user-images.githubusercontent.com/9023041/32708384-925f7b58-c7f7-11e7-83e4-8b18a3c7672f.png)

# Output of commands in RStudio Shell after that reinstall:

![image](https://user-images.githubusercontent.com/9023041/32708444-0da5cfc4-c7f8-11e7-9479-8fd770c5b8b0.png)

*****

@jennybc The "Use Git from Windows Command Prompt" was not selected by default during my first installation. It *was* selected by default only if I tried reinstalling Git after having already installed Git with that option selected (if that makes sense).

**Comment by kevinushey** [comment]
Thanks for following up! I'll have to re-test this -- I am pretty sure in my case that the default behavior in the installer was to select "Use Git from the Command Prompt", but it seems as though that's not true in general. Even so, we should make sure that Git integration works even if that option is not selected.

**Comment by everetr** [comment]
I will make a PR to note this in section 8.1 of the book.

**Comment by jennybc** [comment]
Thanks @everetr!. Should it go in http://happygitwithr.com/install-git.html#windows? Because seems like a fact that is good to know at Git install time.

**Comment by everetr** [comment]
@jennybc You're right. Thanks! Will do.

**Comment by jennybc** [comment]
Closed via 6f8ec675f04aadc026eba242e960409aa12dd86c

**Comment by Igneous01** [comment]
Hi, I just installed git using the exact same settings as OP and I get the same issue:

![image](https://user-images.githubusercontent.com/5571589/33692688-871a3abc-dabc-11e7-88f4-38f8dc3919ee.png)

I tried OPs suggestion and reinstalled (checking use git from windows command prompt option) but it still does not work.

I'm not able to get Git CMD Working either (Although maybe I dont understand the difference between GitCMD and Git Bash):

![image](https://user-images.githubusercontent.com/5571589/33692729-bf786640-dabc-11e7-82f4-a93eb3e5937e.png)

**Comment by melindahiggins2000** [comment]
Thank you all for this issue post. I just ran into this issue on a new Windows computer (had to get my old one repaired), so I'm setting everything back up. I'm glad to get the terminal working so I can teach my students this new awesome feature of RStudio this semester!

**Comment by arnold-c** [comment]
Thanks very much to those who put together this excellent resource. I'm afraid that I, unfortunately, am having issues with Git within RStudio, and was wondering if anyone had a bright idea for the solution? Initially, it was working with my repos, and then one day they started to show the error "RStudio, directory name is invalid".

I am using W10, RStudio 1.1.463, git version 2.19.1.windows.1

I have tried the solutions to previous similar questions (repo paths do not contain non-ASCII char). I can use git with SourceTree, so I don't think it's an issue with my git version.

When I try to commit/pull/push changes it just shows "RStudio, directory name is invalid" and does not allow me to perform any git commands. Also, when I look at the git viewer in RStudio, it is just blank/white, whereas it would usually allow me to stage files etc.

In an attempt to try and fix the issue, I uninstalled and reinstalled git, following the steps above. Whilst git works perfectly well with SourceTree and in the command line, RStudio is not able to find it i.e. in creating a VCS project it returns "Git was not detected on the system path". However, I have placed C:\Program Files\Git\bin\ in the system path via environment variables. I have also set the git executable path in RStudio's global options as C:/Program Files/Git/bin/git.exe.

Any help would be appreciated, as whilst it's not essential, it would be nice to get working again

**Comment by kevinushey** [comment]
@arnold-c are you attempting to use git with a project that's located on a network drive by any chance? (What is the output of `getwd()` in your project?)

**Comment by arnold-c** [comment]
@kevinushey No, it's on my laptop's C: Drive (`getwd()` outputs `C:/Users/owner/Documents/Repos/rti`).

**Comment by everetr** [comment]
@kevinushey Sidenote: Are there known issues with using Git+RStudio on a network drive? Is this Windows-specific?

**Comment by arnold-c** [comment]
@kevinushey Thanks for your help. I've submitted a bug report and referenced this discussion.

**Comment by lunatic-fringe566** [comment]
I'm stuck on a beginner problem of using git please a little Help is really appreciated I'm stuck on this for many days😟. So Whenever I create a new folder (For ex: repo)  and on doing pwd (i.e., repo pwd ) even on changing directory (i.e., cd repo) it still always shows "command not found"  
please help......!!🙏
